digital_land/package/sqlite.py

```diff
@@ -46,18 +49,100 @@ def spatialite(self, path=None):
                     path = "/usr/lib/x86_64-linux-gnu/mod_spatialite.so"
         self._spatialite = path
 
-    def connect(self):
+    def connect(self, optimised=False):
         logging.debug(f"sqlite3 connect {self.path}")
-        self.connection = sqlite3.connect(self.path)
+        self.connection = None
+
+        if optimised:
+            self.connection = sqlite3.connect(self.path, cached_statements=0)
+        else:
+            self.connection = sqlite3.connect(self.path)
 
         if self._spatialite:
             self.connection.enable_load_extension(True)
             self.connection.load_extension(self._spatialite)
 
     def disconnect(self):
         logging.debug("sqlite3 disconnect")
+        if self.cursor:
+            self.cursor.close()
+            self.cursor = None
+
         self.connection.close()
 
+    def create_cursor(self, transactional=False):
+        if self.cursor:
+            self.cursor.close()
+            self.cursor = None
+
+        self.cursor = self.connection.cursor()
+        self.cursor.execute("PRAGMA synchronous = OFF")
+        self.cursor.execute("PRAGMA journal_mode = OFF")
+        if transactional:
+            self.cursor.execute("BEGIN")
```

Comment:
I'm not we need to execute BEGIN here. It makes more sense to me to add the into the functions that are doing the loading rather than here. Also for the batching you've implemented above will it only apply to the first batch? Once the commit is ran?

Also doesn't executemany automatically negate the need for this? As it executes all of the statements in one transaction?

Could we not just leave the insert function the same but run a begin before it then not even need to store anything in a python list? Have you tried this?

Reply:
If we add BEGIN to the loading functions, then they contain SQL and functional code.
I'm trying to centralize the SQL code to one place, to make it easier to look after in future.
I used a Transactional param to satisfy your previous wish to have other clients of SQLite package choose whether they use Transactions or not.
The BEGIN is applied at connection-level, not cursor level. We just need to make sure the 2 separate connections (one for the DDL statements, and the other for the INSERTS) can independently control whether they use transactions or not.
The commit does not cancel the Transaction, closing the current connection does.

Execute many does execute all in one transaction, and _may_ apply one if it's not there.
Never a bad thing to be explicit in our intent, though - especially as we've created a connection and cursor at this point.

Yes I tried that and many other permutations.
All modifications were made incrementally (see Trello ticket comments). Simply using a BEGIN does not provide the speed increase that this version does.